### ISSUE TITLE
`v3.31.0` causes wrong R53 zone to be queried even with `R53_ZONE()` specifying ID

### ISSUE THREAD

**Issue by tjstansell** [StackExchange/dnscontrol#2296]
We started seeing errors where one of our deployment thought it needed to remove hundreds of records.  We use the R53 provider and include `R53_ZONE()` entries in the `D()` blocks to specify which exact zoneid each zone belongs to (some zones have both public and private ones of an identical name).  Our CI system picked up an updated version of dnscontrol and it became clear that when `D('myzone.net!public')` was being evaluated, it's pulling in records from the `D('myzone.net!private')` R53 zone.  There's a TXT record that identifies the zone and the diff clearly showed the problem.  I then went and started downloading previous versions of dnscontrol to identify when the problem arose and the problem exists in `v3.31.0` but not in `v3.30.0`, so we are going to pin our version to `v3.30.0` until this can be resolved.

**Comment by tlimoncelli** [comment]
Thanks for narrowing it down to which version first included the bug.

Can you provide a minimal dnsconfig.js that demonstrates the problem?  (redact any secrets or anything proprietary, of course!).  If you could also attach the output of "dnscontrol print-ir" that would be appreciated.

Does the problem change if you run under diff2?  i.e. `dnscontrol --diff2=true preview` (please test with `=true` and `=false`)

**Comment by tjstansell** [comment]
OK.  I went and created two test route53 zones in a dev account and created a stub config that looks a lot like what we do for our production configs... 

Here's the basic dnscontrol.js config:
```
var DSP_R53 = NewDnsProvider("r53_main");
var REG_NONE = NewRegistrar("none");

D('testzone.net!private', REG_NONE,
        R53_ZONE('Z0376337PMJCCCP1V7UW'),
        DnsProvider(DSP_R53),
        NAMESERVER_TTL('1d'),
        NAMESERVER('ns-1024.awsdns-00.org.'),
        NAMESERVER('ns-0.awsdns-00.com.'),
        NAMESERVER('ns-1536.awsdns-00.co.uk.'),
        NAMESERVER('ns-512.awsdns-00.net.'),
        DefaultTTL('1h')
);
D('testzone.net!public', REG_NONE,
        R53_ZONE('Z01438711DINNG98SHJQ2'),
        DnsProvider(DSP_R53),
        NAMESERVER_TTL('1d'),
        NAMESERVER('ns-1084.awsdns-07.org.'),
        NAMESERVER('ns-184.awsdns-23.com.'),
        NAMESERVER('ns-1710.awsdns-21.co.uk.'),
        NAMESERVER('ns-997.awsdns-60.net.'),
        DefaultTTL('1h')
);

D_EXTEND('testzone.net!public',
    TXT('me',	'public testzone.net'),
    IGNORE_NAME('__nothing__')
);

D_EXTEND('testzone.net!private',
    TXT('me',	'private testzone.net'),
    IGNORE_NAME('__nothing__')
);
```

I did an initial `push` using the known good version ...

```
# dnscontrol version
dnscontrol "3.30.0" ("f700184e3803c9b60cf43dc5960155a39da721cf") built 29 Mar 23 11:01 UTC
# dnscontrol push --config test.js
******************** Domain: testzone.net!private
1 correction (r53_main)
#1:
± MODIFY TXT me.testzone.net: ("private testzone.net" ttl=300) -> ("private testzone.net" ttl=3600)
± MODIFY NS testzone.net: (ns-512.awsdns-00.net. ttl=172800) -> (ns-512.awsdns-00.net. ttl=86400)
± MODIFY NS testzone.net: (ns-1024.awsdns-00.org. ttl=172800) -> (ns-1024.awsdns-00.org. ttl=86400)
± MODIFY NS testzone.net: (ns-0.awsdns-00.com. ttl=172800) -> (ns-0.awsdns-00.com. ttl=86400)
± MODIFY NS testzone.net: (ns-1536.awsdns-00.co.uk. ttl=172800) -> (ns-1536.awsdns-00.co.uk. ttl=86400)

SUCCESS!
******************** Domain: testzone.net!public
1 correction (r53_main)
#1:
± MODIFY TXT me.testzone.net: ("public testzone.net" ttl=300) -> ("public testzone.net" ttl=3600)
± MODIFY NS testzone.net: (ns-997.awsdns-60.net. ttl=172800) -> (ns-997.awsdns-60.net. ttl=86400)
± MODIFY NS testzone.net: (ns-184.awsdns-23.com. ttl=172800) -> (ns-184.awsdns-23.com. ttl=86400)
± MODIFY NS testzone.net: (ns-1084.awsdns-07.org. ttl=172800) -> (ns-1084.awsdns-07.org. ttl=86400)
± MODIFY NS testzone.net: (ns-1710.awsdns-21.co.uk. ttl=172800) -> (ns-1710.awsdns-21.co.uk. ttl=86400)

SUCCESS!
Done. 2 corrections.
```
and verified that things look consistent using this version 
```
# dnscontrol preview --config test.js
******************** Domain: testzone.net!private
******************** Domain: testzone.net!public
Done. 0 corrections.
```
Testing the latest version, you can see it shows the incorrect output...
```
# ./dnscontrol-3.31.2 --diff2=false preview --config test.js
[INFO: Old diff algorithm in use. Please test --diff2 as it will be the default in releases after 2023-05-07. See https://github.com/StackExchange/dnscontrol/issues/2262]
******************** Domain: testzone.net!private
******************** Domain: testzone.net!public
1 correction (r53_main)
#1:
± MODIFY TXT me.testzone.net: ("private testzone.net" ttl=3600) -> ("public testzone.net" ttl=3600)
± MODIFY NS testzone.net: (ns-0.awsdns-00.com. ttl=86400) -> (ns-1084.awsdns-07.org. ttl=86400)
± MODIFY NS testzone.net: (ns-1024.awsdns-00.org. ttl=86400) -> (ns-1710.awsdns-21.co.uk. ttl=86400)
± MODIFY NS testzone.net: (ns-1536.awsdns-00.co.uk. ttl=86400) -> (ns-184.awsdns-23.com. ttl=86400)
± MODIFY NS testzone.net: (ns-512.awsdns-00.net. ttl=86400) -> (ns-997.awsdns-60.net. ttl=86400)

Done. 1 corrections.
# ./dnscontrol-3.31.2 --diff2=true preview --config test.js
[INFO: Diff2 algorithm in use.]
******************** Domain: testzone.net!private
******************** Domain: testzone.net!public
1 correction (r53_main)
#1: ± MODIFY testzone.net NS (ns-0.awsdns-00.com. ttl=86400) -> (ns-1084.awsdns-07.org. ttl=86400)
± MODIFY testzone.net NS (ns-1024.awsdns-00.org. ttl=86400) -> (ns-1710.awsdns-21.co.uk. ttl=86400)
± MODIFY testzone.net NS (ns-1536.awsdns-00.co.uk. ttl=86400) -> (ns-184.awsdns-23.com. ttl=86400)
± MODIFY testzone.net NS (ns-512.awsdns-00.net. ttl=86400) -> (ns-997.awsdns-60.net. ttl=86400)
± MODIFY me.testzone.net TXT ("private testzone.net" ttl=3600) -> ("public testzone.net" ttl=3600)
Done. 1 corrections.
```
and the `print-ir` output that you asked for...
```
# ./dnscontrol-3.31.2 print-ir --config test.js
{"registrars":[{"name":"none","type":"-"}],"dns_providers":[{"name":"r53_main","type":"-"},{"name":"bind","type":"-"}],"domains":[{"name":"testzone.net","registrar":"none","dnsProviders":{"r53_main":-1},"meta":{"ns_ttl":"86400","zone_id":"Z0376337PMJCCCP1V7UW"},"records":[{"type":"TXT","name":"me","ttl":3600,"txtstrings":["private testzone.net"],"target":"private testzone.net"}],"nameservers":[{"name":"ns-1024.awsdns-00.org"},{"name":"ns-0.awsdns-00.com"},{"name":"ns-1536.awsdns-00.co.uk"},{"name":"ns-512.awsdns-00.net"}],"ignored_names":[{"pattern":"__nothing__","types":"*"}],"unmanaged":[{"label_pattern":"__nothing__","rType_pattern":"*"}]},{"name":"testzone.net","registrar":"none","dnsProviders":{"r53_main":-1},"meta":{"ns_ttl":"86400","zone_id":"Z01438711DINNG98SHJQ2"},"records":[{"type":"TXT","name":"me","ttl":3600,"txtstrings":["public testzone.net"],"target":"public testzone.net"}],"nameservers":[{"name":"ns-1084.awsdns-07.org"},{"name":"ns-184.awsdns-23.com"},{"name":"ns-1710.awsdns-21.co.uk"},{"name":"ns-997.awsdns-60.net"}],"ignored_names":[{"pattern":"__nothing__","types":"*"}],"unmanaged":[{"label_pattern":"__nothing__","rType_pattern":"*"}]}]}
```

**Comment by tjstansell** [comment]
Oh, and FWIW, `dnscontrol print-ir` using the `3.30.0` version produces the exact same output.

**Comment by tlimoncelli** [comment]
Very helpful!  Thanks!  I'm traveling but I should be able to take a look on Sunday.

**Comment by tlimoncelli** [comment]
I think I see the problem.  When you have domain!public and domain!private, the GetZoneRecords() call just gets "domain" and doesn't know which (public or private) domain is intended. As a result, you have a 50/50 chance of getting the wrong domain records.  This affects all providers.

I have a few ideas on how to fix this. (and I'm surprised nobody has noticed this so far!)

**Comment by tjstansell** [comment]
Thanks. But isn't that the point of `R53_ZONE()`? Why is that zone ID not being used? Just a regression? Are there tests that should have caught this?

**Comment by tlimoncelli** [comment]
It's a regression.  Sadly there aren't tests that cover that situation and I'm not sure how to add a test that would cover it.  That part of the code isn't written in a way that permits unit tests :-(  (which is probably why this bug cropped up!)

**Comment by tlimoncelli** [comment]
I found the problem!  A recent refactoring made the zone id data inaccessible to the provider.  Sadly the fix this I had to change the signature of a function that appears in many places, thus the PR is quite large.

Give the code on branch `tlim_b2296_r53zoneid` a try.

To compile it:

```
git clone https://github.com/StackExchange/dnscontrol.git
cd dnscontrol
git checkout tlim_b2296_r53zoneid
go install
```

This will put a new binary in `~/bin`

Test that binary and let me know the results.
